### DISCUSSION TITLE
C# Bindings

### DISCUSSION THREAD

**Discussion by GrauBlitz** [godot-jolt/godot-jolt#243] [Questions] [answered]
Hello, first things first: Thank you for the work on this extension, as Godot Physics are totally useless currently and it seems that this will be the case for quite some time, i had also looked into Jolt and then found your Repo.

My question is: Will this be also have C# Bindings and will be usable in C# Scripts? Or is this automatically generated for Godot Extensions? The docs on this is still really spare, so I'm not sure.

Thanks!

**Comment by mihe** [comment, marked as answer]
Hello!

The first phase of this extension will purely be an implementation of the `PhysicsServer3DExtension` interface that Godot exposes, with nothing else added to it. So you would just use the same node types that you use today with Godot Physics, like `RigidBody3D`, `PhysicsDirectSpaceState3D` and whatever else, meaning no extra bindings are needed and everything should just work in C# in the same way it does without this extension.

With that said, there _might_ come a second phase where I do start adding more stuff on top, and I must admit I don't actually know how you would expose a `GDCLASS` from an extension to C#. When you're in GDScript the editor seems to pick up on extension-based classes like `JoltPhysicsDirectSpaceState3D` automatically, so the script server seems to be aware of its presence, but somehow that needs to be compiled into a .NET assembly for C# to pick up on it I guess.

Anyway, that's for much later, if ever. C# support is a must-have for me, as I tend to prefer it over GDScript, but I haven't prioritized bindings in this first phase for the reasons mentioned above.

**Reply by GrauBlitz** [reply]
Thanks for the answer, yeah C# is also my "only option" when coding in Godot. Another question: It seems that almost all basic properties e.g. Bias, Softness, Relaxation, Restitution, Damping, Force Limit, ... are marked as "indicates that it's blocked or incompatible with Jolt in some way." -> "Seemingly not supported by Jolt.", so how will it be possible to configure springs and limits etc., as these parameters are needed as far as i know to tune the physics interactions. Example: Vehicle suspension where I need different spring settings depending of the type of the vehicle and it's mass. Is it possible/planned (maybe in phase two) to implement some Jolt specific settings? (Sorry if this is a dumb question, as i am just starting to look into the engine and in parallel i'm trying to integrate your extension for some basic tests in my project)

**Reply by mihe** [reply]
> so how will it be possible to configure springs and limits etc., as these parameters are needed as far as i know to tune the physics interactions

The short and honest answer is that I don't really know at this time. This is something I've been meaning to look into a bit more before I start rolling out beta releases, which should hopefully happen within the next couple of weeks. The problem is that many of Godot's joint parameters don't seem to line up with what Jolt expects, and I suspect that's just how it's gonna have to be. I understand that this might be a bummer to people who have come to expect them and rely on them, but I figured there could still be some use for these joints even if they don't support springs or soft limits.

> Is it possible/planned (maybe in phase two) to implement some Jolt specific settings?

Exposing more of Jolt's constraints/joints, and exposing the existing ones with a more suitable interface, would be my main reason to move into this second phase. That would let me expose stuff like position motors for some new `Jolt6DOFJoint` node, which could be configured with spring frequency/damping, and could be used for things like vehicle suspension. I can't make any promises as for when (or even if) that will happen though.

**Reply by mihe** [reply]
Just as a brief update, I've now more or less committed to adding new node types for Jolt's constraints/joints as part of the work for `v1.0.0`. You can track this work on the new [project board](https://github.com/orgs/godot-jolt/projects/1) that I created earlier today.

As for C#, the situation is [worse than I thought](https://github.com/godot-jolt/godot-jolt/issues/270). Not only can you not use extension-specific classes from C#, you currently can't even use some of the regular Godot classes like `PhysicsServer3D`, `PhysicsDirectBodyState3D` or `PhysicsDirectSpaceState3D`. As mentioned in the issue, I've made a pull request to address this, but it hasn't really moved forward since then.
